pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -620,69 +622,76 @@ private void completeCommittingSegments() {
-     * TODO Modify numPartitions field in segment metadata and re-write it in propertystore.
-     * The numPartitions field in the metadata is used by SegmentCompletionManager
-     */
-    for (int i = 0; i < nSegments; i++) {
-      final LLCSegmentName segmentName = segmentNames.get(i);
-      if (segmentName.getPartitionId() == curPartition) {
-        final String curSegmentNameStr = segmentName.getSegmentName();
-        if (!segmentNamesIS.contains(curSegmentNameStr)) {
-          LOGGER.info("{}:Repairing segment for partition {}. Segment {} not found in idealstate", realtimeTableName, curPartition,
-              curSegmentNameStr);
-
-          final ZNRecord partitionAssignment = getKafkaPartitionAssignment(realtimeTableName);
-          List<String> newInstances = partitionAssignment.getListField(Integer.toString(curPartition));
-          LOGGER.info("{}: Assigning segment {} to {}", realtimeTableName, curSegmentNameStr, newInstances);
-          // TODO Re-write num-partitions in metadata if needed.
-
-          String prevSegmentNameStr = null;
-          // If there was a prev segment in the same partition, then we need to fix it to be ONLINE.
-          if (i < nSegments-1) {
-            LLCSegmentName prevSegmentName = segmentNames.get(i+1);
-            if (prevSegmentName.getPartitionId() == segmentName.getPartitionId()) {
-              prevSegmentNameStr = prevSegmentName.getSegmentName();
-            }
-          }
-          updateHelixIdealState(realtimeTableName, newInstances, prevSegmentNameStr, curSegmentNameStr);
-          // Skip all other segments in this partition.
+    final ZNRecord partitionAssignment = getKafkaPartitionAssignment(realtimeTableName);
+    for (Map.Entry<Integer, MinMaxPriorityQueue<LLCSegmentName>> entry : partitionToLatestSegments.entrySet()) {
+      final LLCSegmentName segmentName = entry.getValue().pollFirst();
+      final String segmentId = segmentName.getSegmentName();
+      final int partitionId = entry.getKey();
+      if (!segmentNamesIS.contains(segmentId)) {
+        LOGGER.info("{}:Repairing segment for partition {}. Segment {} not found in idealstate", realtimeTableName,
+            partitionId, segmentId);
+
+        List<String> newInstances = partitionAssignment.getListField(Integer.toString(partitionId));
+        LOGGER.info("{}: Assigning segment {} to {}", realtimeTableName, segmentId, newInstances);
+        // TODO Re-write num-partitions in metadata if needed.
+        // If there was a prev segment in the same partition, then we need to fix it to be ONLINE.
+        LLCSegmentName prevSegmentName = entry.getValue().pollLast();
+        String prevSegmentNameStr = null;
+        if (prevSegmentName != null) {
+          prevSegmentNameStr = prevSegmentName.getSegmentName();
         }
-        curPartition--;
+        updateHelixIdealState(realtimeTableName, newInstances, prevSegmentNameStr, segmentId);
```

Comment:
Can we rename `updateHelixIdealState` to better describe what it does? I assume it's doing something close to `addNewSegmentToIdealState`